MoreLinq.NoConflictGenerator/Program.cs

```diff
@@ -0,0 +1,225 @@
+                            break;
+                        case "":
+                            continue;
+                        default:
+                            dir = arg.Current[0] != '-'
+                                ? arg.Current
+                                : throw new Exception("Invalid argument: " + arg.Current);
+                            break;
+                    }
+                }
+            }
+
+            Func<string, bool> PredicateFromPattern(string pattern, bool @default) =>
+                string.IsNullOrEmpty(pattern)
+                ? delegate { return @default; }
+                : new Func<string, bool>(new Regex(pattern).IsMatch);
+
+            var includePredicate = PredicateFromPattern(includePattern, true);
+            var excludePredicate = PredicateFromPattern(excludePattern, false);
+
+            var thisAssemblyName = typeof(Program).GetTypeInfo().Assembly.GetName();
+
+            var q =
+                from fp in Directory.EnumerateFiles(dir, "*.cs")
+                where !excludePredicate(fp) && includePredicate(fp)
+                select new
+                {
+                    SourcePath = fp,
+                    Methods =
+                        from cd in
+                            CSharpSyntaxTree
+                                .ParseText(File.ReadAllText(fp), CSharpParseOptions.Default.WithPreprocessorSymbols("MORELINQ"))
+                                .GetRoot()
+                                .SyntaxTree
+                                .GetCompilationUnitRoot()
+                                .DescendantNodes().OfType<ClassDeclarationSyntax>()
+                        where (string) cd.Identifier.Value == "MoreEnumerable"
+                        from md in cd.DescendantNodes().OfType<MethodDeclarationSyntax>()
+                        let mn = (string) md.Identifier.Value
+                        where md.ParameterList.Parameters.Count > 0
+                        && md.ParameterList.Parameters.First().Modifiers.Any(m => (string)m.Value == "this")
+                        && md.Modifiers.Any(m => (string)m.Value == "public")
+                        && md.AttributeLists.SelectMany(al => al.Attributes).All(a => a.Name.ToString() != "Obsolete")
+                        select md
+                };
+
+            var indent = new string('\x20', 4);
+            var indent2 = indent + indent;
+            var indent3 = indent2 + indent;
+
+            var baseImports = new []
+            {
+                "System",
+                "System.CodeDom.Compiler",
+                "System.Collections.Generic",
+            };
+
+            var imports =
+                from ns in baseImports.Concat(usings)
+                select indent + $"using {ns};";
+
+            var classes =
+                from f in q
+                from md in f.Methods
+                group md by (string) md.Identifier.Value into g
+                select new
```

Comment:
What makes you think it's not deterministic already? Grouping is guaranteed to be done in a deterministic order per [the documentation](https://docs.microsoft.com/en-gb/dotnet/api/system.linq.enumerable.groupby?view=netframework-4.5.1#System_Linq_Enumerable_GroupBy__2_System_Collections_Generic_IEnumerable___0__System_Func___0___1__):

> The [`IGrouping<TKey,TElement>`](https://docs.microsoft.com/en-gb/dotnet/api/system.linq.igrouping-2?view=netframework-4.5.1) objects are yielded in an order based on the order of the elements in `source` that produced the first key of each [`IGrouping<TKey,TElement>`](https://docs.microsoft.com/en-gb/dotnet/api/system.linq.igrouping-2?view=netframework-4.5.1). Elements in a grouping are yielded in the order they appear in `source`.

Reply:
`Directory.EnumerateFiles(dir, "*.cs")` is not deterministic (nor should it be), and I don't know if the roslyn syntax tree iterates deterministically (although this sounds like it should be the case, otherwise noops could not be implemented). However, even if the files were sorted, I think it is a good idea to sort on method names, to prevent spurious changes due to formatting changes (ie, the output should be indifferent to the order of declaration).

Reply:
Ah, I see your point now. I thought you were more specifically concerned about the grouping since you called out `g` specifically. I agree about making the enumeration of files deterministic through sorting.

> if the roslyn syntax tree iterates deterministically

I'm quite sure this is deterministic.